### PR TITLE
feat(android): idling resource refactoring

### PR DESCRIPTION
**Pull request by gosha212**
## Description

- This pull request addresses the issue described here: #4175 

In this pull request, I have made refactored the idling resources handling and creation as a first step in new arch implementation

---

> _For features/enhancements:_
 - [ ] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [ ] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ReactNativeIdlingResources.kt [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import org.joor.Reflect
+
+
+class ReactNativeIdlingResources(
+    private val reactContext: ReactContext,
+    private var launchArgs: LaunchArgs,
+    private val idlingResourcesFactory: DetoxIdlingResourceFactory = DetoxIdlingResourceFactory(reactContext)
+) {
+    companion object {
+        const val LOG_TAG = "DetoxRNIdleRes"
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ReactNativeIdlingResources.kt:25]: Isn't it better to use a file-scoped `private const val` for this?

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/animations/AnimatedModuleIdlingResource.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Hooks up to React Native internals to monitor the state of the animations.
+ *
+ */
+class AnimatedModuleIdlingResource(private val reactContext: ReactContext) : DetoxIdlingResource(),
+    Choreographer.FrameCallback {
+
+    override fun getName(): String {
+        return AnimatedModuleIdlingResource::class.java.name
+    }
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/animations/AnimatedModuleIdlingResource.kt:26]:
```suggestion
    override fun getName() = AnimatedModuleIdlingResource::class.java.name
```

etc. etc.
I think you should do this all across the board, it makes things much more concise and far less java-like

File: detox/android/detox/src/full/java/com/wix/detox/espresso/EspressoDetox.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ public static void setURLBlacklist(final ArrayList<String> urls) {
         UIThread.postSync(new Runnable() {
             @Override
             public void run() {
-                NetworkIdlingResource.setURLBlacklist(urls);
+                NetworkIdlingResource.Companion.setURLBlacklist(urls);
```
Comment (d4vidi) [inline detox/android/detox/src/full/java/com/wix/detox/espresso/EspressoDetox.java:123]: Can't we use `@JvmStatic` to clean this up to:
```suggestion
                NetworkIdlingResource.setURLBlacklist(urls);
```

File: detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/timers/TimersIdlingResourceTest.kt [hunk 4]
```diff
@@ -0,0 +1,212 @@
+package com.wix.detox.reactnative.idlingresources.timers
```
Comment (d4vidi) [inline detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/timers/TimersIdlingResourceTest.kt:1]: 🔥

File: detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResourceTest.kt [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.never
+import org.mockito.kotlin.times
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.whenever
+import org.robolectric.RobolectricTestRunner
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+
+private class AsyncStorageModuleStub : NativeModule {
```
Comment (d4vidi) [inline detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResourceTest.kt:24]: Get we get rid of Spek altogether, then?

**Review by d4vidi — APPROVED** [review]
I did a semi-shallow review; Approve but with comments/suggestions (to pick up later) 🙏🏻
Anyways, awesome cleanup!

File: detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResourceTest.kt (hunk 5 above)
Comment (gosha212) [reply detox/android/detox/src/testFull/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResourceTest.kt:24]: Still have a few spec tests

**Pull request merged**